[WIP] nixos: add environment.gnome3.enableExtensions

bjornfor opened the pull request:
###### Motivation for this change
I'd like to enable/disable GNOME3 extensions from configuration.nix. However, when testing this I learned that services.xserver.desktopManager.gnome3.extraGSettingsOverrides, which this is built upon, only provides _default_ values for the system. Users are still able to change settings. And once they have made a change, any change to the system default will be ignored. This is marked WIP since I'd like for this setting to be truly declarative. The first 4 commits would be nice to merge anyway, so one can play with this locally at least.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

bjornfor committed:
nixos/gnome3: add environment.gnome3.enableExtensions option

This option takes a list of packages / GNOME extensions that will be
(declaratively) enabled. The default value is null, meaning the
extensions are managed imperatively (backwards compatible).

hedning commented:
dconf settings are stored per user (under `~/.config/dconf`), so it would be hard to enforce fully I think.

But something like running `dconf reset /org/gnome/shell/enabled-extensions` as the user on startup might work OK though?

bjornfor commented:
> dconf settings are stored per user (under ~/.config/dconf), so it would be hard to enforce fully I think.

According to [this](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/desktop_migration_and_administration_guide/custom-default-values-system-settings#lock-down-specific-settings), one should be able to create lock files that prevent users from doing modifications. In my last fixup commit you can see I try to use it, but it didn't work for me.

Your "dconf reset ..." command sounds good. What setting in NixOS is best for running commands at user login time? (I'd like something shell agnostic.)

jtojnar commented:
See also https://github.com/NixOS/nixpkgs/issues/54150

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/recommended-way-to-add-gnome-extensions/5589/3

bjornfor committed:
fixup! nixos/gnome3: add environment.gnome3.enableExtensions option

Try to add add a lock file, to prevent users from modifying / overriding
the system (default) settings. However, it doesn't work.

bbigras commented:
Any progress on this?

bjornfor commented:
Not from my side.

bbigras commented:
What is needed to move this PR forward?

Does #54150 needs to be done first?

bjornfor commented:
> Does #54150 needs to be done first?

Yep, that's what I'm thinking.

Alizter commented:
Any news on this?

bjornfor commented:
> Any news on this?

No, https://github.com/NixOS/nixpkgs/issues/54150 still needs to land first.

linsui commented:
You can use the dconf module now.

bjornfor commented:
> You can use the dconf module now.

Right, there was another PR (not linked here) that fixed it! Cool!

I'll put this back on my TODO list.

ulysses4ever commented:
https://github.com/NixOS/nixpkgs/issues/54150 has ben merged. Anything else left before this one can land?

bjornfor commented:
> #54150 has ben merged. Anything else left before this one can land?

This PR needs a rewrite. It's not high priority for me, as I'm using home-manager for my gnome extensions. Someone please make new PR on top of #54150.

WillsterJohnson commented:
> > #54150 has ben merged. Anything else left before this one can land?
> 
> This PR needs a rewrite. It's not high priority for me, as I'm using home-manager for my gnome extensions. Someone please make new PR on top of #54150.

How are you doing this via home-manager? I looked at [your repo](https://github.com/bjornfor/nixos-config/) but I see it's been a while since that mirror got an update

bjornfor commented:
> How are you doing this via home-manager?

Using the dconf module. For example:

```nix
{ pkgs, lib, ... }:

let
  # Make widgets smaller to not clip other widgets (default: 100).
  graphWidth = 95;
in
{
  dconf.settings."org/gnome/shell".enabled-extensions = [
    pkgs.gnomeExtensions.system-monitor-next.extensionUuid
  ];

  dconf.settings = {
    "org/gnome/shell/extensions/system-monitor" = {
      # Highlight IO wait with dark red
      cpu-iowait-color = "#a40000ff";

      disk-display = true;
      # In the default color scheme it's impossible to separate reads and
      # writes.
      disk-read-color = "#00ff00ff"; # bright green
      disk-write-color = "#ff0000ff"; # bright red

      cpu-graph-width = graphWidth;
      memory-graph-width = graphWidth;
      net-graph-width = graphWidth;
      disk-graph-width = graphWidth;
    };
  };
}
```

> I looked at [your repo](https://github.com/bjornfor/nixos-config/) but I see it's been a while since that mirror got an update

Yes, I intentionally stopped updating it.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/changing-gdm-gsettings-declaratively/49579/8

bjornfor commented:
> Since we can enable extensions that way (and through [home-manager](https://github.com/NixOS/nixpkgs/pull/46433#issuecomment-2440041987)), is this PR still relevant? @bjornfor @jtojnar

I still think it's valuable to have a simple interface. Why not turn your code above into a NixOS module with `environment.gnom3.enableExtensions = [ ... ]` interface?

jtojnar commented:
You would need to interpolate it with `${pkgs.gnomeExtensions.system-monitor.extensionUuid}`, otherwise the attribute set would contain something like the following:

```nix
{
  pkgs = {
    gnomeExtensions = {
      system-monitor = { extensionUuid = { key = "value"; }; };
      user-themes = { extensionUuid = { key = "value"; }; };
    };
  };
}
```

But I do not think there is a relation between the UUID and schema id for most extensions.

---

Also, I generally prefer generic options instead of creating specific options (in line with [RFC42](https://github.com/NixOS/rfcs/pull/42)). `enableExtensions` would probably be fine since we can just redirect it to `programs.gsettings.settings."org.gnome.shell".enabled-extensions` + `programs.gsettings.schemaPackages` once that is implemented. It would be harder to do that for random extensions without somehow keeping track of the UUID → package mapping.
